### ISSUE TITLE
Give permissions, commands and more ranges of effect

### ISSUE THREAD

**Issue by DevJake** [DevJake/SaltBot-2.0#25]
When using the bot, a system of 'ranges' should be available. These ranges allow for a permission or command to have a wider - or smaller - area of effect. Ranges come in four types:
- Global
- Guild-wide
- TextChannel-wide
- User-wide

### Examples
For each range type, an example has been given. The examples are in terms of permissions. 
#### Global
A permission with a global range will be applied to _every user_ that shares a guild with the bot. As a result, this range should (**must**) be limited to bot administrators. 
#### Guild-wide
A permission with a guild-wide range will be applied to _every user_ in the Guild that the range was applied in. 
#### TextChannel-wide
A permission with a TextChannel-wide range will be applied to _every user_ in the TextChannel that the range was applied in. 
#### User-wide
A permission with a User-wide range will be applied to a _single user_. This is the most common range used, and allows for the finest level of permission control and restriction. 

### Permissions for ranges
Due to the imminent amount of power a range can offer, each range should also require a specific permission to use it. A simplistic system would be to append an applicable range to each base permission. Take, for example, a command that requires the following permission:

> salt.commands.currency.givemoney

If an individual is given this permission, they become capable of giving money to any user, from a virtual, infinitely-sized bank of money. With each command, a default range is declared. For this command, the default range is **User-wide**; the command can only be applied to a single user at a time. 

By applying a **_Guild-wide_** range, the individual can instead increase the currency of all guild members, simultaneously. Of course, this can be a very powerful tool, so a permission should be required to allow usage of such a range;

> salt.command.currency.givemoney.range_guild

Anyone with this permission will now be able to use the command on a Guild-wide range. 

#### Permissions and global ranges
The global range can be very dangerous if not properly managed. If a permission were to be added or removed globally, it can easily cause mass havoc. As a result, the global permission should only be allowed to be applied to a predefined list of users, namely bot administrators. 

##### Global ranges and self-hosted bots
Bots that are being self-hosted should allow for a custom list of users allowed to use the global range. As a note, it is **incredibly important** that this list can only be edited when self-hosting. 

## Summary
To summarise ranges:
- A permission can be ranged to provide access to all members of a Guild or TextChannel, or applied to an individual user. 
- A command can be ranged to allow for wider coverage in its actions. Commands are ranged by their _permissions_

Ranges aim to allow for improvements to the efficiency of permission management, as well as improvements to command usages. Increasing the balance of three users would only take a few seconds, therefore disregarding the requirement for ranges, but increasing the balances for _2000_ users will take a very long time. As a result, ranges allow for more efficient usage of the bot, as well as for finer tuning of configurations.

Ranges are considered an **advanced** feature, and should be used with exercised caution.
